dtm0/recovery.c

```diff
@@ -0,0 +1,2170 @@
+		       "are no remote DTM0 services.");
+
+	m0_tl_for(rfom, &m->rm_rfoms, rf) {
+		M0_LOG(M0_DEBUG,
+		       "id=" FID_F ", is_volatile=%d, "
+		       "is_local=%d, state=%s, got_eol=%d => %d",
+		       FID_P(&rf->rf_tgt_svc),
+		       (int) rf->rf_is_volatile,
+		       (int) rf->rf_is_local,
+		       m0_ha_state2str(rf->rf_last_known_ha_state),
+		       (int) rf->rf_last_known_eol,
+		       (int) was_log_replayed(rf));
+	} m0_tl_endfor;
+}
+
+static bool is_local_recovery_completed(struct m0_dtm0_recovery_machine *m)
+{
+	M0_ENTRY();
+	rec_cond_trace(m);
+	return M0_RC(m0_tl_exists(rfom, r, &m->rm_rfoms, !r->rf_is_local) &&
+		m0_tl_forall(rfom, r, &m->rm_rfoms, was_log_replayed(r)));
+}
+
+static void remote_state_update(struct recovery_fom    *rf,
+				const struct eolq_item *item)
+{
+	M0_ENTRY("rf=%p, " FID_F " state: %s, eol: %d",
+		 rf, FID_P(&rf->rf_tgt_svc),
+		 m0_ha_state2str(rf->rf_last_known_ha_state),
+		 (int) rf->rf_last_known_eol);
+
+	switch (item->ei_type) {
+	case EIT_HA:
+		M0_LOG(M0_DEBUG, "new_state=%s",
+		       m0_ha_state2str(item->ei_ha_state));
+		rf->rf_last_known_ha_state = item->ei_ha_state;
+		/* Clear the EOL flag if the remote is dead. */
+		if (M0_IN(item->ei_ha_state, (M0_NC_TRANSIENT, M0_NC_FAILED)))
+			rf->rf_last_known_eol = false;
+		break;
+	case EIT_EOL:
+		M0_LOG(M0_DEBUG, "new_eol=1");
+		rf->rf_last_known_eol = true;
+		break;
+	default:
+		M0_IMPOSSIBLE("Wrong eolq item type %d?", item->ei_type);
+		break;
+	}
+
+	M0_LEAVE();
+}
+
+static void local_recovery_fom_coro(struct m0_fom *fom)
+{
+	struct recovery_fom *rf = M0_AMB(rf, fom, rf_base);
+	struct recovery_fom *remote_rf;
+
+	M0_CO_REENTER(CO(fom),
+		      struct m0_be_op      eolq_op;
+		      bool                 eoq;
+		      bool                 recovered;
+		      struct eolq_item     item;
+		      enum m0_ha_obj_state state;);
+
+	F(eoq) = false;
+	/*
+	 * A DTM0 service without persistent storage does not need
+	 * REDOs.
+	 * mkfs does not require DTM0 support as well.
+	 */
+	F(recovered) = rf->rf_is_volatile;
+
+	/* Wait until the moment where we should start recovery. */
+	do {
+		M0_CO_FUN(CO(fom), heq_await(fom, &F(state), &F(eoq)));
+		if (F(eoq))
+			goto out;
+	} while (F(state) != M0_NC_DTM_RECOVERING);
+
+	while (!F(recovered)) {
+		M0_CO_FUN(CO(fom), eolq_await(fom, &F(item)));
+		M0_ASSERT(F(item).ei_type != EIT_END);
+
+		recovery_machine_lock(rf->rf_m);
+		remote_rf = recovery_fom_by_svc_find(rf->rf_m,
+						     &F(item).ei_source);
+		if (remote_rf == NULL) {
+			/* XXX: machine is stopping? */
+			recovery_machine_unlock(rf->rf_m);
+			break;
+		}
+		M0_ASSERT(remote_rf != rf);
+		remote_state_update(remote_rf, &F(item));
+
+		F(recovered) = is_local_recovery_completed(rf->rf_m);
+		recovery_machine_unlock(rf->rf_m);
+	}
+
+	M0_BE_QUEUE__FINISH(&rf->rf_eolq, typeof(F(item)));
+
+	/*
+	 * Emit "RECOVERED". It shall cause HA to tell us to transit from
+	 * RECOVERING to ONLINE.
+	 */
+	if (F(recovered)) {
```

Comment:
Is it possible to have 'F(recovered) == false'?
If we 'break' from Line No. 1872, that will be.
If that happens, what should we do?

Reply:
3:Working on it.  Adding comments in the code.

Reply:
Fixed/merged.